pinot-core/src/main/java/com/linkedin/pinot/core/operator/aggregation/function/AggregationFunction.java

```diff
@@ -17,87 +17,81 @@
 
 import com.linkedin.pinot.core.operator.aggregation.AggregationResultHolder;
 import com.linkedin.pinot.core.operator.aggregation.groupby.GroupByResultHolder;
-import java.util.List;
+import com.linkedin.pinot.core.operator.docvalsets.ProjectionBlockValSet;
+import java.io.Serializable;
+import javax.annotation.Nonnull;
 
 
 /**
- * Interface for Aggregation functions.
+ * Interface for aggregation functions.
  */
-public interface AggregationFunction {
+// TODO: revisit whether Serializable is necessary.
+public interface AggregationFunction<AggregationResult extends Serializable, FinalResult> {
 
-  enum ResultDataType {
-    LONG,
-    DOUBLE,
-    AVERAGE_PAIR,
-    MINMAXRANGE_PAIR,
-    DISTINCTCOUNT_SET,
-    DISTINCTCOUNTHLL_HYPERLOGLOG,
-    HLL_PREAGGREGATED,
-    PERCENTILE_LIST,
-    PERCENTILEEST_QUANTILEDIGEST
-  }
+  /**
+   * Accept an aggregation function visitor to visit.
+   */
+  void accept(@Nonnull AggregationFunctionVisitorBase visitor);
+
+  /**
+   * Get aggregation result holder.
+   */
+  @Nonnull
+  AggregationResultHolder getAggregationResultHolder();
 
-  void accept(AggregationFunctionVisitorBase visitor);
+  /**
+   * Get group-by result holder with the given initial capacity, max capacity and trim size.
+   */
+  @Nonnull
+  GroupByResultHolder getGroupByResultHolder(int initialCapacity, int maxCapacity, int trimSize);
+
+  /**
+   * Perform aggregation on the given projection block value sets.
+   */
+  void aggregate(int length, @Nonnull AggregationResultHolder aggregationResultHolder,
+      @Nonnull ProjectionBlockValSet... projectionBlockValSets);
```

Comment:
Does this need to be ProjectionBlockValSet, or can it be BlockValSet? If latter, please use that so we have more flexibility for change in future. If not, this will be an issue when supporting UDF's, ie when we are aggregation on the output of UDF